src/builtins/path.cpp

```diff
@@ -0,0 +1,892 @@
+        if (pos == wcstring::npos || pos == 0) {
+            return none();
+        }
+
+        // Convert pos back to what it would be in the original path.
+        return pos + path.size() - filename.size();
+}
+
+static int path_extension(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        if (!pos) {
+            // If there is no extension the extension is empty.
+            // This is unambiguous because we include the ".".
+            path_out(streams, opts, L"");
+            continue;
+        }
+
+        wcstring ext = arg->substr(*pos);
+        if (opts.quiet && !ext.empty()) {
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_change_extension(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 1, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        wcstring ext;
+        if (!pos) {
+            ext = *arg;
+        } else {
+            ext = arg->substr(0, *pos);
+        }
+
+        // Only add on the extension "." if we have something.
+        // That way specifying an empty extension strips it.
+        if (*opts.arg1) {
+            if (opts.arg1[0] != L'.') {
+                ext.push_back(L'.');
+            }
+            ext.append(opts.arg1);
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_resolve(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
+            // something that does.
+            wcstring next = *arg;
+            // First add $PWD if we're relative
+            if (!next.empty() && next[0] != L'/') {
+                next = wgetcwd() + L"/" + next;
```

Comment:
This should use `parser.vars().get_pwd_slash()` for two reasons:

1. It allows sanely handling cases where `getcwd` fails (most importantly if $PWD has been unlinked)
2. It plays better with concurrent execution, where each thread might have its own CWD

Be aware that `get_pwd_slash()` may have symlinks.

Also note the `path_apply_working_directory` function, this may be as simple as

    wcstring next = path_apply_working_directory(*arg, parser.vars.get_pwd_slash());

Reply:
As the kids don't say: Donespacito.